### PR TITLE
Fully merge celery in

### PR DESCRIPTION
**Pull request by MichaelDeng03**
- Remove and replace LlmMixin with AbstractOperator
- Add response_format: str to ConcreteChatCompletion for client-side model validation via get_response
- Implement _do_work_celery + add support for celery in communicative_dehallucination

### CODE REVIEW

**Comment by kgang** [comment]
awesome work

**Comment by kgang** [comment]
I'm getting an error for this, can you take a look
```
import sys
LOCAL_CONCRETE_PATH = '/Users/kentgang/git/concrete/'
sys.path = [LOCAL_CONCRETE_PATH] + sys.path

from concrete import clients, operators, messages
c = {'openai': clients.OpenAIClient(temperature=0)}
dev = operators.Developer(clients=c)
resp = dev.chat(
    "Deploy the provided project to AWS. The project directory is: `print('hello world!')`",
    tools=[tools.AwsTool],
    response_format=messages.Tool
)  # Errors
```
Error
```
BadRequestError: Error code: 400 - {'error': {'message': "Invalid schema for response_format 'Tool': In context=('properties', 'tool_parameters'), 'default' is not permitted.", 'type': 'invalid_request_error', 'param': 'response_format', 'code': None}}
```

**Comment by MichaelDeng03** [comment]
yep

**Comment by MichaelDeng03** [comment]
Okay, I know that you can't have optional fields with openai structured outputs

**Comment by MichaelDeng03** [comment]
I've changed Tool to this
class Tool(Message):
    tool_name: str = Field(description="Name of the tool")
    tool_function: str = Field(description="Command to call the tool")
    tool_parameters: list[str] = Field(description="Parameters to pass into the tool function call.")
    tool_keyword_parameters: dict[str, str] = Field(
        description="Parameters specificed by keyword to pass into the tool function call."
    )

however, I am getting another bug
BadRequestError: Error code: 400 - {'error': {'message': "Invalid schema for response_format 'Tool': In context=(), 'required' is required to be supplied and to be an array including every key in properties. Extra required key 'tool_keyword_parameters' supplied.", 'type': 'invalid_request_error', 'param': 'response_format', 'code': None}}

**Comment by kgang** [comment]
That was the original error I saw. Didn't even realize the error changed when I added a default lol

**Comment by MichaelDeng03** [comment]
lol

**Comment by MichaelDeng03** [comment]
I think the issue is the dict[str, str], kinda stupid that openai can't parse it, but you probably have to define another pydantic model
class kw_params:
  k: str
  v: str

**Comment by kgang** [comment]
does it work if you try it?

**Comment by MichaelDeng03** [comment]
I didn't try making another model, but I removed dict[str, str] and replaced it with str

**Comment by MichaelDeng03** [comment]
All of the other message models work, so I can't imagine it's anything else

**Comment by kgang** [comment]
Can you try it and look at the documentation and see if they say dicts like that are not allowed

**Comment by MichaelDeng03** [comment]
Yeah I found this while I was debugging

[Slack File: image.png](https://files.slack.com/files-pri/T07DSK6AKDJ-F07KD7KUXPY/image.png)

**Comment by MichaelDeng03** [comment]
<https://platform.openai.com/docs/guides/structured-outputs/how-to-use>

**Comment by MichaelDeng03** [comment]
I figured that dict would be converted to Object, but ig not

**Comment by kgang** [comment]
that's wild

**Comment by kgang** [comment]
let's remove `tool_keyword_parameters` entirely

**Comment by MichaelDeng03** [comment]
okay

**Comment by kgang** [comment]
Do you mind also documenting what's pydantic primitives are allowed under `ConcreteModel` or `Message`?

**Comment by kgang** [comment]
I was just messing around with trying to throw an error if something is marked Optional, but I ran into similar problems as when I was looking at the nesting problem

**Comment by kgang** [comment]
Just thinking aloud - it'd be nice to tailor the json response format to have dedicated slots for each required param on a tool by tool basis. Something tricky though is that a tool might have several functions on it with different parameters

**Comment by MichaelDeng03** [comment]
That would be nice, maybe we could dynamically create the message based on tool function definitions

File: concrete/abstract.py [hunk 1]
```diff
@@ -31,31 +31,45 @@ def abstract_operation(operation: Operation, clients: dict[str, OpenAIClientMode
...
 
 
 class MetaAbstractOperator(type):
+    """
+    This metaclass automatically creates a '_delay' version for each method in the class, allowing these methods to be executed asynchronously using Celery workers.
+
+    Classes that use this metaclass can call the asynchronous variant of a method like `some_instance.some_method.delay()`.
+
+    Note that .delay() returns a Celery AsyncResult object, which can be retrieved using .get() to get a ConcreteChatCompletion object.
+    """  # noqa E501
```
Comment (kgang) [inline concrete/abstract.py:60]: nice

File: Makefile [hunk 2]
```diff
@@ -67,7 +67,10 @@ aws_ecr_push_demo: aws_ecr_login
 	docker push 008971649127.dkr.ecr.us-east-1.amazonaws.com/webapp-demo:latest
 
 rabbitmq:
+	docker rm -f rabbitmq || true
```
Comment (kgang) [inline Makefile:70]: add rmq and celery commands to helloworld celery example

maybe combine the rabbitmq and celery commands unless there's a need to run them separately

File: concrete/abstract.py [hunk 3]
```diff
@@ -3,19 +3,19 @@
...
 from .clients import CLIClient, OpenAIClient
 from .models.clients import ConcreteChatCompletion, OpenAIClientModel
+from .models.messages import Message, TextMessage
 from .models.operations import Operation
-from .models.responses import Response, TextResponse
 from .tools import MetaTool
 
-
 # TODO replace OpenAIClientModel with GenericClientModel
+
```
Comment (kgang) [inline concrete/abstract.py:17]: Do you agree with this TODO? I think we should take it out

File: concrete/models/messages.py [hunk 4]
```diff
@@ -11,53 +11,73 @@ class MyClass(BaseModel):
     response = self.client.beta.chat.completions.parse(messages=messages, temperature=self.OPENAI_TEMPERATURE, **kwargs)
     message = response.choices[0].message
     message_formatted: MyClass = message.parsed
+    
+    
+# Allowed primitives
+https://platform.openai.com/docs/guides/structured-outputs/supported-schemas
+The following types are supported for Structured Outputs:
+String, Number, Boolean, Integer, Object, Array, Enum, anyOf
```
Comment (kgang) [inline concrete/models/messages.py:19]: Can you please add that `Optional` is not allowed? thanks

File: concrete/models/messages.py [hunk 5]
```diff
@@ -11,53 +11,73 @@ class MyClass(BaseModel):
...
 from typing import List, Optional
 
 from pydantic import Field
 
-from .base import ConcreteBaseModel, KombuMixin
+from .base import ConcreteModel, KombuMixin
 
+# Tracks all message types created as a sub class of Message
+# Keys are not type sensitive
+RESPONSE_REGISTRY = {}
```
Comment (kgang) [inline concrete/models/messages.py:30]:
```suggestion
RESPONSE_REGISTRY: dict[str, "Message"] = {}
```
sorry to ask you to improve my code lmfao

File: concrete/models/messages.py [hunk 6]
```diff
@@ -11,53 +11,73 @@ class MyClass(BaseModel):
...
+    tool_parameters: list[str] = Field(description="Parameters to pass into the tool function call.")
 
 
-class Tools(Response):
+class Tools(Message):
     tools: List[Tool] = Field(description="List of tools")
 
 
-class ProjectFile(Response, KombuMixin):
+class ProjectFile(Message, KombuMixin):
```
Comment (kgang) [inline concrete/models/messages.py:60]:
```suggestion
# N.B. KombuMixin must be added to each leaf child node class due to serializer registration
class ProjectFile(Message, KombuMixin):
```

File: concrete/orchestrator.py [hunk 7]
```diff
@@ -103,42 +101,59 @@ async def _do_work_plain(self) -> AsyncGenerator[tuple[str, str], None]:
                 tools=[AwsTool],
                 response_format=Tool,
             )
-            invoke_tool(**deploy_tool_call.__dict__)  # nosec
+            invoke_tool(**deploy_tool_call.dict())  # nosec
```
Comment (kgang) [inline concrete/orchestrator.py:104]:
```suggestion
            invoke_tool(**deploy_tool_call.dict())
```

File: concrete/orchestrator.py [hunk 8]
```diff
@@ -103,42 +101,59 @@ async def _do_work_plain(self) -> AsyncGenerator[tuple[str, str], None]:
...
 
         self.update(status=ProjectStatus.FINISHED)
         yield Developer.__name__, str(files)
 
     # TODO: implement using Celery task calls
     async def _do_work_celery(self) -> AsyncGenerator[tuple[str, str], None]:
-        components = self.exec.plan_components(self.starting_prompt, response_format=PlannedComponents).components
+        # components = self.exec.plan_components(self.starting_prompt, response_format=PlannedComponents).components
```
Comment (kgang) [inline concrete/orchestrator.py:111]:
```suggestion
```

File: concrete/orchestrator.py [hunk 9]
```diff
@@ -103,42 +101,59 @@ async def _do_work_plain(self) -> AsyncGenerator[tuple[str, str], None]:
...
+            .get()
+            .get_response()
         )
 
         if self.deploy:
-            if self.aws is None:
-                raise ValueError("Cannot deploy without AWSAgent")
-            self.aws.deploy(files, self.uuid)
+            # TODO Use an actual DB instead of emulating one with a dictionary
+            # TODO Figure something out safer than eval
```
Comment (kgang) [inline concrete/orchestrator.py:146]:
```suggestion
```

**Review by kgang — COMMENTED** [review]
almost there

File: concrete/abstract.py (hunk 3 above)
Comment (MichaelDeng03) [reply concrete/abstract.py:17]: I think it's something we should do eventually. Definitely not a first priority, but there's no reason to believe we won't be using other LLM clients in the future.

Comment (kgang) [reply concrete/abstract.py:17]: Ok [](edited-with-Axolo)

File: concrete/models/messages.py (hunk 5 above)
Comment (MichaelDeng03) [reply concrete/models/messages.py:30]: lolol, I think this was my code actually, I yoinked the registry idea from your tool registry though

Comment (kgang) [reply concrete/models/messages.py:30]: lol

**Comment by MichaelDeng03** [comment]
make helloworld_celery is broken rn

File: Makefile (hunk 2 above)
Comment (MichaelDeng03) [reply Makefile:70]: Kinda hard to combine helloworld with celery/rabbitmq commands - it takes a sec for them to startup and it doesn't seem like they're waiting for server startup correctly.

Comment (kgang) [reply Makefile:71]: Hmm just put a sleep 5-30 seconds? [](edited-with-Axolo)

**Review by kgang — APPROVED** [review]
LGTM

**Review by kgang — APPROVED** [review]
Double LGTM

**Comment by kgang** [comment]
Maybe give it a better title

**Pull request merged**